packages/backend-common/src/reading/FetchUrlReader.ts

```diff
@@ -57,15 +57,34 @@ export class FetchUrlReader implements UrlReader {
   };
 
   async read(url: string): Promise<Buffer> {
+    const response = await this.readUrl(url);
+    return response.buffer();
+  }
+
+  async readUrl(
+    url: string,
+    options?: ReadUrlOptions,
+  ): Promise<ReadUrlResponse> {
     let response: Response;
     try {
-      response = await fetch(url);
+      response = await fetch(url, {
+        headers: {
+          ...(options?.etag && { 'If-None-Match': options.etag }),
+        },
+      });
     } catch (e) {
       throw new Error(`Unable to read ${url}, ${e}`);
     }
 
+    if (response.status === 304) {
+      throw new NotModifiedError();
+    }
+
     if (response.ok) {
-      return Buffer.from(await response.text());
+      return {
+        buffer: async () => Buffer.from(await response.text()),
+        etag: response.headers.get('ETag') ?? undefined,
```

Comment:
Can we lowercase these headers and it all still work fine or are they parsed and cap'd by node?

Reply:
We could. Headers are case-insensitive by the RFC and they always converted to lowercase in node internally. I started with lowercase names but I used the pattern that was established in the [GithubUrlReader](https://github.com/backstage/backstage/blob/2bdf2dd143c3a0de4cff03b6b4a549c52401e544/packages/backend-common/src/reading/GithubUrlReader.ts#L114). Not sure why @Rugvip decided to do that. At least `ETag` is the official name in the ([RFC7232#2.3](https://datatracker.ietf.org/doc/html/rfc7232#section-2.3)).

Reply:
meh. ok. if there's prior art for it let's keep it this way. 😅

Reply:
I find that header literals are generally title cased unless accessed via a case sensitive API, which no header API should be 😅. A bit nicer to read and often how they are sent over the wire as well